atomicapp/providers/openshift.py

```diff
@@ -36,91 +35,205 @@ class OpenShiftProvider(Provider):
-                    logger.info("Processing template")
-                    artifact = self._processTemplate(artifact_path)
-                kube_order[data["kind"].lower()] = artifact
+                # kind has to be specified in artifact
+                if "kind" not in data.keys():
+                    raise ProviderFailedException(
+                        "Error processing %s artifact. There is no kind" %
+                        artifact)
+
+                kind = data["kind"].lower()
+
+                # process templates
+                if kind == "template":
+                    processed_objects = self._process_template(data)
+                    # add all processed object to artifacts dict
+                    for obj in processed_objects:
+                        obj_kind = obj["kind"].lower()
+                        if obj_kind not in self.openshift_artifacts.keys():
+                            self.openshift_artifacts[obj_kind] = []
+                        self.openshift_artifacts[obj_kind].append(obj)
+                    continue
+
+                # add parsed artifact to dict
+                if kind not in self.openshift_artifacts.keys():
+                    self.openshift_artifacts[kind] = []
+                self.openshift_artifacts[kind].append(data)
+
+    def _process_template(self, template):
+        """ process tempalate
+
+        return list of templated object from template
+        """
+        logger.debug("processing template: %s", template)
+        url = self._get_url(self._get_namespace(template), 'processedtemplates')
+        res = requests.post(url, json=template, verify=self.ssl_verify)
+        if res.status_code == 201:
+            logger.info("template proccessed %s", template['metadata']['name'])
+            logger.debug("processed template %s", res.json())
+            return res.json()['objects']
+        else:
+            msg = "%s %s" % (res.status_code, res.content)
+            logger.error(msg)
+            raise ProviderFailedException(msg)
+
+    def _kind_to_resource(self, kind):
+        """ converts kind to resource name
+
+        same logic as in k8s.io/kubernetes/pkg/api/meta/restmapper.go
+        KindToResource
+        """
+        singular = kind.lower()
+        if singular.endswith("status"):
+            plular = singular + "es"
+        else:
+            if singular[-1] == "s":
+                plular = singular
+            elif singular[-1] == "y":
+                plular = singular.rstrip("y") + "ies"
             else:
-                raise ProviderFailedException("Malformed artifact file")
+                plular = singular + "s"
+        return plular
+
+    def _get_url(self, namespace, kind, name=None):
+        """ return url for given kind
+
+        return either openshift or kubernetes url
+        depending on what kind is passed
+        """
+        url = None
+
+        resource = self._kind_to_resource(kind)
+
```

Comment:
Explain the url you are piecing together in all of these following commands. Showing what the url is composed of and one resulting url would be nice.